federation/pkg/federation-controller/replicaset/pod_helper.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replicaset
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/kubernetes/federation/pkg/federation-controller/util"
+	api_v1 "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+type PodAnalysisResult struct {
+	// Total number of pods created.
+	Total int
+	// Number of pods that are running and ready.
+	RunningAndReady int
+	// Number of pods that have been in unschedulable state for a while.
+	Unschedulable int
+
+	// TODO: Handle other scenarios like pod waiting too long for scheduler etc.
+}
+
+const (
+	UnschedulableThreshold = 60 * time.Second
+)
+
+// A function that for calculates how many pods from the list are in one of
+// the meaninfgul (from the replica set perspective) state. This function is
+// a temporary workaround against the current lack of ownerRef in pods.
+func AnalysePods(replicaSet *v1beta1.ReplicaSet, allPods []util.FederatedObject, currentTime time.Time) (map[string]PodAnalysisResult, error) {
+
+	selector, err := labelSelectorAsSelector(replicaSet.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid selector: %v", err)
+
+	}
+	result := make(map[string]PodAnalysisResult)
+
+	for _, fedObject := range allPods {
+		pod, isPod := fedObject.Object.(*api_v1.Pod)
+		if !isPod {
+			return nil, fmt.Errorf("invalid arg content - not a *pod")
+		}
+
+		if !selector.Empty() && selector.Matches(labels.Set(pod.Labels)) {
```

Comment:
@quinton-hoole - when do you expect someone having 100 clusters with 500.000 pods each? I wouldn't expect any time in the next 2 years. I agree it's not optimal, but we should solve the same problem in pure kubernetes first and just migrate the solution here.